vms/platformvm/docs/validators_versioning.md

```diff
@@ -0,0 +1,113 @@
+| SubnetID + Reverse_Height + NodeID | validator.BlsKey bytes or nil |
+
+Note that:
+
+1. `BLS Public Key` diffs have the same keys as `Weight` diffs. This implies that the same ordering is guaranteed.
+2. Value is either validator `BLS Public Key` bytes or an empty byte slice, as illustrated in previous section.
+
+## Validators diff usage in rebuilding validators state
+
+Now let's see how diffs are used to rebuild the validator set at a given height. The procedure varies slightly between Primary Network and Subnet validator, so we'll describe them separately.
+We assume that the reader knows that, as of the Cortina fork, every Subnet validator must also be a Primary Network validator.
+
+### Primary network validator set rebuild
+
+If the P-Chain's current height is `T` and we want to retrieve the Primary Network validators at height `H < T`. We proceed as follows:
+
+1. We retrieve the Primary Network validator set at current height `T`. This is the base state on top of which diffs will be applied.
+2. We apply weight diffs first. Specifically:
+   -  `Weight` diff iteration starts from the top-most height smaller or equal to `T`. Remember that entries heights do not need to be contiguous, so the iteration starts from the highest height smaller or equal to `T`, in case `T` has not diff entry.
+   -  Since `Weight` diffs are forward-looking, each diff is applied in reverse. A validator's weight is decreased if `state.ValidatorWeightDiff.Decrease` is `false` and it is increased if it is `true`.
+   - We take care of adding or removing a validator from the base set based on its weight. Whenever a validator weight, following diff application, becomes zero, we drop it; conversely whenever we encounter a diff increasing weight for a currently-non-existing validator, we add the validator to the base set.
+   -  The iteration stops at the first height smaller or equal to `H+1`. Note that a `Weight` diff stored at height `K` holds the content to turn validator state at height `K-1` into validator state at height `K`. So to get validator state at height `K` we must apply diff content at height `K+1`.
+3. Once all `Weight` diffs have been applied the resulting validator set will contain all Primary Network validators active at height `H` and only those. However these validators have the `BLS Public Keys` registered at height `T` which may be still set wrong: a validator may have restaked between height `H` and `T`, and may have registered a different `BLS Public Key` at the two height, or not registered any `BLS Public Key` at all at either height. We solve this by applying `BLS Public Key` diffs to the vallidator set:
+   - Once again we iterate `BLS Public Key` diffs from the top-most height smaller or equal to `T` till the first height smaller or equal to `H+1`.
+   - Since `BLS Public Key` diff are *backward-looking*, we simply nil the BLS key when diff is nil and we restore the BLS Key when it is specified in the diff. 
```

Comment:
```suggestion
   - Since `BLS Public Key` diffs are *backward-looking*, we simply nil the BLS key when diff is nil and we restore the BLS Key when it is specified in the diff. 
```